### DISCUSSION TITLE
Cloning an Entry

### DISCUSSION THREAD

**Discussion by PaulMcF1987** [Budibase/budibase#5828] [Help]
On the Edit Entry screen `tablename/:id` I have edited the form to give it steps.
The first step of the form is intended to allow me to edit the data in the row.
The second step is intended to allow me to clone the data and save a new row. However, this process is proving to be very difficult

I have created a custom Query `CloneQuery`

The clone query looks like:

`INSERT INTO tablea(`
		`JobTitle, Customer, BookedBy, Datebooked, Price,  JobStatus, Cost, JobNumber`
`)`
`VALUES (`
	`{{ Form.Fields.CloneJobtitle }}, {{ Form.Fields.CloneAccounts }}, {{ Form.Fields.CloneBookedby }}, {{ Form.Fields.CloneDateBooked }}, {{ Form.Fields.Cloneprice }}, {{ Form.Fields.CloneStatus }}, {{ Form.Fields.Clonecost }}, {{ Form.Fields.CloneJobno }}`
`);`

The values used are taken from the `Field` value on the input settings.

The save button onclick values have been set to
`Validate Form`
`Execute Query (Datasource - MYSQL, Query - CloneQuery)`
`Navigate to (back to table)`

Something that should be very simple has taken me hours and still no further forward. 
After pressing save, it tables me back to the table but there are no new jobs to show.
The table is as it was before I tried cloning

**Comment by melohagan** [comment]
Hi @PaulMcF1987 

I think you may have misunderstood the bindings. You have to pass the form fields through to your query.

So your Bindings section of the Query should look like:
![Screenshot 2022-05-11 at 14 29 02](https://user-images.githubusercontent.com/101575380/167861457-c5924478-37d0-4de7-bd74-cd9f3dd392cf.png)

And then your use of the query from the design section:
![Screenshot 2022-05-11 at 14 30 49](https://user-images.githubusercontent.com/101575380/167861799-aaca8309-8395-451a-9ba9-9bd377b2a14a.png)

There's more on that here: https://docs.budibase.com/docs/custom-queries

Hopefully that makes sense!

**Reply by melohagan** [reply]
Also make sure function is set to create:

![Screenshot 2022-05-11 at 14 34 34](https://user-images.githubusercontent.com/101575380/167862518-a6b41495-dede-4c7e-8244-f7aa0ea06be0.png)

**Comment by PaulMcF1987** [comment]
@melohagan I have never used bindings before. Any time I have needed to populate a table from Form Data, a Query similar to the above has been enough.
Why are bindings needed?

**Reply by melohagan** [reply]
When you use a Query as a datasource, it's able to grab the columns from the schema e.g.
![Screenshot 2022-05-11 at 14 39 51](https://user-images.githubusercontent.com/101575380/167863570-c68c3d95-7c48-4360-bf39-f0387e6272a6.png)

Whenever you're doing an Insert, there is no schema as such. Your SQL could be doing many inserts to many tables.

**Reply by melohagan** [reply]
I just noticed there is also a **Duplicate Row** action which may be useful:
![Screenshot 2022-05-11 at 14 46 09](https://user-images.githubusercontent.com/101575380/167864933-2807518d-eed5-4b78-adf4-fe16987c6d3b.png)

**Comment by PaulMcF1987** [comment]
@melohagan Duplicate wont work as I dont want to duplicate the whole row. There are times where jobs are similar but there are certain cells which need to be chanced. So Ideally I want to be able to take the data onto a form, change the data that needs changed and then save as a new entry

**Reply by melohagan** [reply]
What you can do then is click the *Add Column* button. This will allow you to override the values you don't want to duplicate.

**Comment by PaulMcF1987** [comment]
@melohagan I dont understand how that would achieve the outcome?
Would that not then duplicate the form and leave empty values which I would then need to edit to fill in the values?

I dont understand wht Budibase does not allow me just to write a query which will execute the command.
I have achieved this on websites I have built, I have use other low-code platforms and all of these platforms allow this. Budibase seems to want to complicate things

**Comment by PaulMcF1987** [comment]
@melohagan the other issue I get is the Field Groups which are set, it seems to automatically either set the entries as Long Text or Number

There are a lot of fields in my form so I tried changing all of the Long Text fields to Text Fields... However in the Text Entry settings, when I type the column name, it tells me The field you are setting is the wrong data type for this component

The Datatype was set to text... I then changed it to TinyText to see if this made a difference but no joy.

I decided to give BudiBase a try to see if it would make things easier but Ill be honest, it has proven to be a nightmare since the get go

**Comment by PaulMcF1987** [comment]
Ah well, game over... In trying to sort an issue I have deleted the full screen that I was working on, there doesnt appear to be any way to revert back
I had an element highlighted, I chose delete and it deleted the full screen rather than the selected element

**Reply by melohagan** [reply]
Undo functionality is in the roadmap: https://github.com/Budibase/budibase/issues/3550

We're always looking to improve the user experience, so you could give Budibase another try at a later point. Hopefully then this free service will be more up to your standards. 

The release changes are outlined each month here: https://github.com/Budibase/budibase/discussions/categories/changelog 

If you are looking for more personal and extensive support, you could try contacting sales: https://budibase.com/pricing/

Feel free to raise bugs and suggest features!

**Comment by PaulMcF1987** [comment]
@melohagan  Its not to do with my standards, its to do with being fit for purpose.

For example, can you or anyone else explain why... when the datatype in my database is set to `text - 10 characters long` and I chose the `Text input` for this function, it gives me the message  `The field you are setting is the wrong data type for this component`
The only input type it will accept is `long text`
If I was to have a form full of Long Text inputs the form would be massive as there are around 40 inputs on the form

**Comment by PaulMcF1987** [comment]
@melohagan I meant to reply and say, I found the problem with the above.

If Budibase creates and formats the DB table then the form inouts it choses are exactly as you would expect.

I have found the problem is when importing a ready made table, Budibase does not seem to recognise the Datatype.

Is this a bug? or is it something that the Budibase team are aware of?

P.S, sorry if I seemed arsey above, stress levels were high lol

**Reply by melohagan** [reply]
No worries :)

When you say importing a ready made table, do you mean fetching tables from a MySQL datasource? e.g.
![Screenshot 2022-05-18 at 14 56 43](https://user-images.githubusercontent.com/101575380/169058405-ec8fea00-14d2-46d9-be73-6c959db4f098.png)

I haven't been able to find an issue myself. If you can replicate it, it would be very helpful if you can raise a bug with the steps to reproduce.

**Reply by PaulMcF1987** [reply]
I have decided for what I am doing now I will create the table within BudiBase, so not sure how I would replicate it.
But essentially all I done was 
  - Add Screen
  - Autogenerated Screen
  - Chose the database I was wanting to populate from

I do have one question... I am not overly comfortable at the minute with the set up on BudiBase...

The app I am making is so that I can record jobs, I am looking to link the specific job to a customer from another table by selecting the customer from a dropdown

Usually I would populate the dropdown options using an SQL Query however, there does not seem to be the option to do this in BudiBase.

I am a bit unsure about the relationship Selector... And am weary incase I chose the wrong option ie 
`many x - many y`
`one x - many y`
`one y - many x`

Could you point me in the direction of how this works? as I have never used this sort of set up.

I do more web development and have studied software development some years ago but have never really used it.

**Reply by melohagan** [reply]
If you intend to use relationships, I recommend sticking with MySQL over the internal BudibaseDB (because it's NoSQL and it has a significant drawback in that you can't filter by relationship). 

It seems to me what you would want would be `one customer - many jobs`. 

Then you could use the Relationship Picker in your 'record job' screen to select the single customer. 
![Screenshot 2022-05-18 at 16 22 57](https://user-images.githubusercontent.com/101575380/169079665-7b2af227-fc87-484d-b73c-1828fccd6dc9.png)

**Reply by PaulMcF1987** [reply]
@melohagan
When I try to select one customer - many jobs I get an error - Error Saving Column I have tried 3-4 times, however when I try one job- many customers
It saves the column no problem

**Reply by PaulMcF1987** [reply]
It appears that as with the above, It does not like me using an existing database. when I created a new database through Budibase with the same info, it worked fine

**Reply by PaulMcF1987** [reply]
@melohagan 
Sorry, I am bombarding with questions but the lack of apparent ability to use queries has me a bit stumped.

Where before I would have had a dropdown to give the status of a job 

    -Not Completed
    -In Progress
    -completed

I would then have 3 pages with tables for each running a query `WHERE Status = `In Progress` for example

Also
I have it set up on the other system so that if I view a customer, I can view the customers details as well as 2 tables
1 table which shows all of the jobs I have done for the customer
the other showing the jobs which are still to be invoiced.

I cannot figure out a way to achieve this with Budibase as I cannot see an are where I can write queries.

I have tried adding another data provider and setting a defined filter but cant seem to get the result needed. I cant find a way to filter so that only jobs for the selected customer are showing

**Reply by melohagan** [reply]
@PaulMcF1987 here is a guide for displaying related data in one screen for the internal BudibaseDB: https://docs.budibase.com/docs/fetching-data#displaying-relational-data

For external datasources you can use custom queries (see my first post above)

**Reply by PaulMcF1987** [reply]
Thanks @melohagan 

I meant to remove this last night as I found the option. To be honest, Budibase has really good functionality, It just seems a bit more complicated that some of the others I have used but I suppose that is something that will sort otself out the more I use it and the more comfortable I get.

-------------------------------------------------------------------------------------------------------------------------------------------------
I feel like a pest though as I have another question...

As I have probably mentioned... I aim to have 3 screens, one to view the details of the job, another to edit and another to clone...

The view details screen is made up of `paragraph` blocks and for the most part this works and shows the required information. 

However, where the information is pulled from a relationship column it displays things a bit differently

Where I could expect the output to say `Customer 1`
The actual output is `[{"Primary Display":"Customer 1","%5B1%5D"}]`

Is there a way to remove this clutter and only display the customer's name? I have searched the Docs but im not really sure what Im looking for,

**Reply by melohagan** [reply]
You're welcome! There can be a bit of a learning curve with Budibase, but the benefit is the flexibility in my opinion. 

The relationship column contains an array of all of the related rows; even if it is a single link. 

The *primaryDisplay* field shows the text that is used as the display column (https://docs.budibase.com/docs/budibasedb#selecting-the-display-column)
It appears highlighted in grey within the tables e.g.
<img width="145" alt="Screenshot 2022-05-19 at 13 27 54" src="https://user-images.githubusercontent.com/101575380/169293081-4f20af9f-000b-4738-bd04-d577a0e06768.png">

The *_id* field is the guid for the relationship. 

To only display the customer name, you would enter the following: `{{ <name of relationship column>.0.Primary Display }}`
Grab the primary display of the first relationship. This can be assumed if you know the relationship is One -> One/Many

**Reply by PaulMcF1987** [reply]
Sorry, I dont follow,

What I would usually enter would be `{{ Repeater.RecordJob.Customer }}`

So I have tried entering `{{ Repeater.RecordJob.Customer.0.Primary Display }}` bit it tells me that ``Current Handlebars syntax is invalid, please check the guide [here](https://handlebarsjs.com/guide/) for more details.

I have also tried `{{ Repeater.RecordJob.Customer.0.PrimaryDisplay }}` (without the space) but still no luck

**Reply by melohagan** [reply]
Sorry, this is the format here: `{{ New Repeater.Product.Manufacturer.0.primaryDisplay }}` - lowercase 'p' in primary. 

I was thrown off a little by this `[{"Primary Display":"Customer 1","%5B1%5D"}]` but the field is *primaryDisplay*

**Reply by PaulMcF1987** [reply]
@melohagan 

Based on the above, I was wondering...

If I have chosen 3 entries via a multi-option dropdown (A, B, C). If I place them into a paragraph element it will display:
[{"primaryDisplay":"A","_id":"%5B1%5D"},{"primaryDisplay":"B","_id":"%5B3%5D"},{"primaryDisplay":"C","_id":"%5B4%5D"}]`

If I then add `.0.primaryDisplay` it will display as `A` with no C or D

I then tried
`{{ Repeater.table.Selection.0.primaryDisplay  }}{{ Repeater.table.Selection.1.primaryDisplay  }}{{ Repeater.table.Selection.2primaryDisplay  }}`
and it displayed`ABC`

I tried putting a `, ` between the handlebars and I got the desired outcome of `A, B, C` but then when I removed C, it displayed `A, B, ,`

My questoin is
Is there a way to show all options, separated by commas or equivalent without having excess commas at the end or not enough commas to cover all the selections?

**Reply by melohagan** [reply]
Hi @PaulMcF1987 

Yes you can do that with a mapper. I believe the handlebars expression is: `{{ pluck Repeater.table.Selection 'primaryDisplay' }}`
